### DISCUSSION TITLE
Add 'Did You Mean' Hints for Deprecated or Renamed Shader Terms in Godot

### DISCUSSION THREAD

**Discussion by DustStock** [godotengine/godot-proposals#10406] [Editor]
I am new to shaders. When I am copying code to test out, I frequently get error messages saying "such and such is invalid".

For several of these, when I google the problem, the term just has a different name in Godot or Godot updated to a new term.

For instance

CAMERA_MATRIX rename to INV_VIEW_MATRIX

All such terms, should have a "did you mean" hint.

This should apply to any term, whether used in other shader languages, or in old versions of Godot, that have a strict 1 to 1 replacement.

This would be incredibly helpful while I am learning shaders. It would allow me to focus on more complex issues and not looking up vocab when there is a clear equivalent.

GDScript occaisionally gives "did you mean..." hints, the shader editor should as well.

I imagine that experienced shader devs coming from other environments would also get a significant boost from this.

**Comment by Calinou** [comment]
- Related to https://github.com/godotengine/godot-proposals/issues/3902.

The implementation in GDScript can't be used as-is in the shader language, so this would need to be reimplemented from scratch in the shader compiler.

Right now, a few outdated names are detected using special cases, but there's no universal system for this in place:

https://github.com/godotengine/godot/blob/88f3b5f9d52f740b24fabfb8bc01b8b7026ba279/servers/rendering/shader_language.cpp#L5668-L5677

**Reply by DustStock** [reply]
It looks like we could replace the code inside that if block with a map lookup that matches terms to replacements.

Then define that map, somewhere, and populate it with terms.

That seems like it would be easily maintained and updated.

But I am unsure of the project norms and not great with cpp. Is there a particular map implementation Godot uses? Is there a preference of where the map would be defined.

**Reply by Calinou** [reply]
> But I am unsure of the project norms and not great with cpp. Is there a particular map implementation Godot uses? Is there a preference of where the map would be defined.

See [C++ usage guidelines](https://docs.godotengine.org/en/latest/contributing/development/cpp_usage_guidelines.html). In Godot 4.x, you can use the HashMap type (probably faster as it's typed) or good old Dictionary.
